Add `matrix_nginx_proxy_ssl_config`

py-crash opened the pull request:
A new variable `matrix_nginx_proxy_ssl_config`  is created to create an _easy switch_ for the SSL configuration of the server, based on [Mozilla's Server TLS recomendations](https://wiki.mozilla.org/Security/Server_Side_TLS#Recommended_configurations).

Also a "Custom" mode with more granular control is added to the playbook.

This also fixes #697, changing the default ciphers.

All changes where properly documented and added to the CHANGELOG.

spantaleev commented on CHANGELOG.md:
```diff
@@ -1,3 +1,15 @@
+
+## New SSL Configuration
+
+A new variable `matrix_nginx_proxy_ssl_config`, bringing some breaking changes. The default configuration is set to `"Intermadiate"`.
```
Typo here. `Intermadiate` -> `Intermediate`.

spantaleev commented on roles/matrix-nginx-proxy/defaults/main.yml:
```diff
@@ -217,9 +217,35 @@ matrix_nginx_proxy_proxy_domain_additional_server_configuration_blocks: []
+
+# Specifies which *SSL protocols* to use when serving all the various vhosts.
+# This option is ignored except you specify "Custom" in "matrix_nginx_proxy_ssl_config"
 matrix_nginx_proxy_ssl_protocols: "TLSv1.2 TLSv1.3"
```
I guess you can make this variable always get used and not only get used when it's custom.

I'd propose something like this:

```yaml
matrix_nginx_proxy_ssl_preset: intermediate

matrix_nginx_proxy_ssl_protocols: "{{ matrix_nginx_proxy_ssl_presets[matrix_nginx_proxy_ssl_preset]['protocols'] }}"

matrix_nginx_proxy_ssl_ciphers: "{{ matrix_nginx_proxy_ssl_presets[matrix_nginx_proxy_ssl_preset]['ciphers'] }}"

matrix_nginx_proxy_ssl_presets:
  intermediate:
    protocols: TLSv1.2 TLSv1.3
    ciphers: ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
  modern:
    protocols: TLSv1.3
    ciphers: ''
  old:
    protocols: TLSv1 TLSv1.1 TLSv1.2 TLSv1.3
    ciphers: ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA
  custom:
    protocols: "{{ matrix_nginx_proxy_ssl_custom_protocols }}"
    ciphers: "{{ matrix_nginx_proxy_ssl_custom_ciphers }}"

matrix_nginx_proxy_ssl_custom_protocols: ''
matrix_nginx_proxy_ssl_custom_ciphers: ''
```

Then, our configuration can just use `matrix_nginx_proxy_ssl_protocols` and `matrix_nginx_proxy_ssl_ciphers` without caring about the "preset" type, etc.

---------

To use the "custom" preset would be something like this:

```yaml
matrix_nginx_proxy_ssl_preset: custom
matrix_nginx_proxy_ssl_custom_protocols: 'something'
matrix_nginx_proxy_ssl_custom_ciphers: 'something'
```

-----------

Well, **another thing we could do** is to not have a `custom` preset (defined in `matrix_nginx_proxy_ssl_presets`) at all and just make people "go custom" like this:

```yaml
matrix_nginx_proxy_ssl_protocols: 'something'
matrix_nginx_proxy_ssl_ciphers: ''
```

This is probably better even, as it means we are reusing the `matrix_nginx_proxy_ssl_protocols` variable we always had and people who were redefining it would not experience any change at all.

Unless I'm missing something, this seems like the better solution.

spantaleev commented on roles/matrix-nginx-proxy/templates/nginx/conf.d/matrix-dimension.conf.j2:
```diff
@@ -65,9 +65,27 @@ server {
 	ssl_certificate {{ matrix_ssl_config_dir_path }}/live/{{ matrix_nginx_proxy_proxy_dimension_hostname }}/fullchain.pem;
 	ssl_certificate_key {{ matrix_ssl_config_dir_path }}/live/{{ matrix_nginx_proxy_proxy_dimension_hostname }}/privkey.pem;
+
+	{% if matrix_nginx_proxy_ssl_config == "Modern" %}
```
Assuming you adopt the other proposed changes, this can get simplified down to:

```nginx
ssl_protocols: {{ matrix_nginx_proxy_ssl_protocols }};

{% if matrix_nginx_proxy_ssl_ciphers == '' %}
ssl_prefer_server_ciphers: no;
{% else %}
ssl_prefer_server_ciphers: yes;
ssl_ciphers: {{ matrix_nginx_proxy_ssl_ciphers }};
{% endif %}
```

spantaleev commented on roles/matrix-nginx-proxy/tasks/validate_config.yml:
```diff
@@ -18,3 +18,9 @@
+  fail:
+    msg: >-
+      `matrix_nginx_proxy_ssl_config` needs to be set to a known value.
+  when: "matrix_nginx_proxy_ssl_config not in ['Custom', 'Old', 'Intermediate', 'Modern']"
```
Besides this check, we should probably add another one that verifies if `matrix_nginx_proxy_ssl_custom_protocols` is set when `matrix_nginx_proxy_ssl_preset == 'custom'`.

spantaleev commented on roles/matrix-nginx-proxy/defaults/main.yml:
```diff
@@ -217,9 +217,35 @@ matrix_nginx_proxy_proxy_domain_additional_server_configuration_blocks: []
+# This is based on the Mozilla Server Side TLS Recommended configurations.
+#
+# The posible values are:
+# - "Modern" - For Modern clients that support TLS 1.3, with no need for backwards compatibility
```
Going for all-lowercase would be more consistent with the naming of the values for other variables in the playbook (e.g. `matrix_ssl_retrieval_method: lets-encrypt`, etc.

spantaleev reviewed (changes requested):
Thanks for working on this! I've proposed some changes to make it simpler and more backward compatible

spantaleev commented:
Hello! Not rushing you at all, but I'm wondering if you plan on finishing this PR?

py-crash commented:
Yes, I'll try to finish it next week. I got caught up in the Holidays

py-crash commented:
Hi! I modified the code. In the end I just removed "custom". Now if a variable is set it will use its value instead of the preset. This will make it easier to mix and match and will also give better backward compatibility

spantaleev commented:
Thanks a lot for making SSL configuration changes so much easier! 👍
